### ISSUE TITLE
Optimizer should track parameter names and not id

### ISSUE THREAD

**Issue by nitishgupta** [pytorch/pytorch#1489]
In the optimizer's `param_groups['params']` the order of the parameters (in which they were given to the optimizer's init) matters. 
In load_state_dict the snippet shows this : 
`id_map = {old_id: p for old_id, p in`
`zip(chain(*(g['params'] for g in saved_groups)),
chain(*(g['params'] for g in groups)))}`
`state = {id_map.get(k, k): v for k, v in state_dict['state'].items()}`

If we change the order in which the parameters are given to the optimizer when loading, the code breaks as the `state` dict now incorrectly maps parameters to their states.

Consider model (when using, say, Adam optimizer)

```
class Model(nn.Module):
    def __init__(self):
    super(Model, self).__init__()
    self.p1 = nn.Linear(2,3, False)
    self.p2 = nn.Linear(3,4, False)
```
After saving, if the order in which the parameters are defined in the model changes i.e. if I change the class to have 
```
self.p2 = nn.Linear(3,4, False)
self.p1 = nn.Linear(2,3, False)
```
the loaded optimizer's state for p1 will be mapped to p2 and vice-versa. I tried this and this indeed happens which is wrong and now training cannot proceed (step() will, rightly so, give an error). 
The `nn.Module` class is robust to such behavior as it uses parameter names instead of id ordering.

IMO the optimizer should also use parameter names instead of ids and relying on the ordering in which they are supplied to the optimizer when initializing.

Corresponding [PyTorch-Discuss post](https://discuss.pytorch.org/t/possible-issues-in-optimizer/2685)

@soumith

**Comment by apaszke** [comment]
I don't think we can do anything to solve it -- `torch.optim` is a package completely independent from `torch.nn`. It can be used in code that doesn't even use Modules, so we can't depend on having named parameters, and I want to keep it this way. Any ideas @soumith @colesbury?

**Comment by bhattad2** [comment]
Any update on this? Thanks! @apaszke

**Comment by IssamLaradji** [comment]
This is quite important as relying on the order of parameters can easily break. This is also important in the cases of re-using the optimizer history for training small a subset of the model (while keeping the rest of the model fixed).
What if the optimizer uses the names of the parameters when the `.named_parameters()` dict is passed ? i.e.
` torch.optim.Adam(model.named_parameters(), lr=1e-5) `

**Comment by apaszke** [comment]
The optimizers don't rely on the order itself. They just rely on the references to the parameters staying valid. If you change your model, you should recreate the optimizer. Even saving names doesn't help because:
1. the optimizer still wouldn't have the reference to the module, so it can't even look them up
2. it makes `optim` tied to `nn` which is undesirable 
3. it's still not robust to parameter changes, except if we list all parameters at every iteration, which is quite costly (or resort to more complicated solutions, like the one for caching DataParallel replicas, but I'd rather not do that)

**Comment by lemairecarl** [comment]
No need to tie `nn` and `optim`. Only require that optimizers are given `named_parameters()`. Thus, in `optimizer.load_state_dict()`, we can associate the right momentum cache with the right model parameter. Instead of associating by ordering, we'd now associate using parameter names.

**Comment by lemairecarl** [comment]
Here is a snippet that crashes but I think shouldn't:
```python
import torch
import torch.nn as nn
import torch.optim as optim
import torch.autograd as autograd

def build_model():
    return nn.Sequential(*[nn.Linear(i, i + 1) for i in range(1, 20)], nn.Linear(20, 1))

def fake_training():
    for _ in range(2):
        out = model(autograd.Variable(torch.Tensor([1.0])))
        optimizer.zero_grad()
        out.backward()
        optimizer.step()

model = build_model()
optimizer = optim.Adam(set(model.parameters()))
fake_training()
s = optimizer.state_dict()

model = build_model()
optimizer = optim.Adam(set(model.parameters()))
optimizer.load_state_dict(s)
fake_training()
```
I find it useful to use set subtraction to manipulate parameter groups... My workaround for now is to use  a list comprehension instead of a set subtraction, so the order is preserved.

**Comment by Santosh-Gupta** [comment]
Is there an update to this? I would love to have optimizers map to the variable name, or least be able to figure out which variable in the model it's 'sum' etc is corresponding to.

**Comment by ezyang** [comment]
cc @vincentqb

**Comment by vincentqb** [comment]
Thanks for the example @lemairecarl. When applying set to the parameters, `model.parameters()` is changed, and all information about which values belong to which parameters is lost. Thus, it should not be expected to give the same answer or work. This is documented [here](https://github.com/pytorch/pytorch/blob/master/torch/optim/optimizer.py#L20).

On one hand, I do agree that the original issue, where the order of `p1` and `p2` is changed changes the model, may be unexpected in a language like python, especially that `nn` appears robust to this. On the other hand, for a change in the model like that to happen, a user is modifying the model between loading/saving the optimizer, and one should be expected to create a new optimizer for the new model.

@Santosh-Gupta -- you can use `model.named_parameters()` to get the name and the values

**Comment by janeyx99** [comment]
Haha, just for reference, this has come up in discussion again recently (see my new proposal here: https://dev-discuss.pytorch.org/t/rfc-introducing-fqns-clarity-to-optim-state-dict/1552).
